File: python/semantic_kernel/connectors/ai/open_ai/responses/open_ai_chat_message_content.py
# Copyright (c) Microsoft. All rights reserved.


from typing import TYPE_CHECKING, Any, Dict, List, Optional

from openai.types.chat import ChatCompletion
from openai.types.chat.chat_completion import Choice

from semantic_kernel.models.contents import ChatMessageContent

if TYPE_CHECKING:
    from semantic_kernel.connectors.ai.ai_request_settings import AIRequestSettings


class OpenAIChatMessageContent(ChatMessageContent):
    """A chat response from OpenAI."""

    inner_content: ChatCompletion
    function_call: Optional[Dict[str, Any]] = None
    tool_calls: Optional[List[Dict[str, Any]]] = None

    def __init__(
        self,
        choice: Choice,
        response: ChatCompletion,
        metadata: Dict[str, Any],
        request_settings: "AIRequestSettings",
        function_call: Optional[Dict[str, Any]],
        tool_calls: Optional[List[Dict[str, Any]]],
    ):
        """Initialize a chat response from OpenAI."""
        super().__init__(
            inner_content=response,
            ai_model_id=request_settings.ai_model_id,
            role=choice.message.role,
            content=choice.message.content,
            metadata=metadata,
            function_call=function_call,
            tool_calls=tool_calls,
        )

File: java/aiservices/openai/src/main/java/com/microsoft/semantickernel/aiservices/openai/chatcompletion/OpenAIChatMessageContent.java
// Copyright (c) Microsoft. All rights reserved.
package com.microsoft.semantickernel.aiservices.openai.chatcompletion;

import com.microsoft.semantickernel.orchestration.FunctionResultMetadata;
import com.microsoft.semantickernel.services.chatcompletion.AuthorRole;
import com.microsoft.semantickernel.services.chatcompletion.ChatMessageContent;
import java.nio.charset.Charset;
import java.util.Collections;
import java.util.List;
import javax.annotation.Nullable;

/**
 * Represents the content of a chat message.
 *
 * @param <T> The type of the inner content.
 */
public class OpenAIChatMessageContent<T> extends ChatMessageContent<T> {

    @Nullable
    private final List<OpenAIFunctionToolCall> toolCall;

    /**
     * Creates a new instance of the {@link OpenAIChatMessageContent} class.
     *
     * @param authorRole   The author role that generated the content.
     * @param content      The content.
     * @param modelId      The model id.
     * @param innerContent The inner content.
     * @param encoding     The encoding.
     * @param metadata     The metadata.
     * @param toolCall     The tool call.
     */
    public OpenAIChatMessageContent(
        AuthorRole authorRole,
        String content,
        @Nullable String modelId,
        @Nullable T innerContent,
        @Nullable Charset encoding,
        @Nullable FunctionResultMetadata metadata,
        @Nullable List<OpenAIFunctionToolCall> toolCall) {
        super(authorRole, content, modelId, innerContent, encoding, metadata);

        if (toolCall == null) {
            this.toolCall = null;
        } else {
            this.toolCall = Collections.unmodifiableList(toolCall);
        }
    }

    /**
     * Gets any tool calls requested.
     *
     * @return The tool call.
     */
    @Nullable
    public List<OpenAIFunctionToolCall> getToolCall() {
        return toolCall;
    }
}
